### PR TITLE
resolve/download packages from arbitrary remotes

### PR DESCRIPTION
**Pull request by vankeisb**
This PR allows to use packages not only from github.com, but from arbitrary git+ssh remotes.
It uses a config file that indicates the host(s) of the package(s), and git commands only (no http).
This allows for easy private/corporate package repos.

See #4.

### CODE REVIEW

**Comment by vankeisb** [comment]
I think this one is better. I have tested it using a corporate github entreprise install, and it works, elm-make compiles etc.

File: src/index.js [hunk 1]
```diff
@@ -1,143 +1,270 @@
...
+
+        var repoName = packageName.split('/')[1];
+
+        var host = repoHost(packageName);
+
+        var cmd = 'cd /tmp && rm -rf ' + repoName +
+            ' && git clone --no-checkout --depth 1 --branch ' +
+            ref + ' git@' + host +
+            ':' + packageName + '.git &&' +
+            'cd elm-cassie && git show HEAD:elm-package.json';
```
Comment (gdotdesign) [inline src/index.js]: I think you left a hard coded reference to the repo your testing the code with: `elm-cassie`

**Comment by gdotdesign** [comment]
Thanks for doing this, I really appreciate it! :+1: 

If you are willing to put some more time into it then I have some suggestions:
* Right now as I see it works on only in unix operating systems and I was looking for a way to resolve that. There is are bindings to `libgit2` https://github.com/tojocky/nodegit and there is a package to handle temporary directories https://github.com/raszi/node-tmp.
* This could be the start of a general `git` version handling, we can remove the archive part and just use the cloning strategy (even for normal github repos).
* And since we are cloning repositories we can get rid of the `git ls-remote` command.

I have some questions too:
* Am I right to think that this in theory could work with any git repository? (bitbucket, gitlab, etc..?)
* Did you use some packages or linters to reformat the code? If there will be more contributors I would like to set up some guidelines.

Thanks again :smile:

File: src/index.js (hunk 1 above)
Comment (vankeisb) [reply src/index.js]: Damnit you're right !

**Comment by vankeisb** [comment]
I agree with your suggestions, they all make sense. I had very little time to spend on that to be fully honest, and I just "hacked my way around" (on a sunday night), using your script as a basis. It can definitely be improved. 

You are right : it's supposed to work with any ssh-enabled git server. Even private ones. 

I have used no linter or anything, again, this is hackish and should not be considered as good practise :P

**Comment by gdotdesign** [comment]
Thanks, do you need it released or can you use your fork until I manage to create a proper release implementing things above?

**Comment by vankeisb** [comment]
It would be awesome if you could release it, I didn't dare to ask for it :P

I'm poking around with Elm at the moment, evaluating and gathering facts. But soon will come the time when my co-workers will need to use the code... and they'll obviously need to build. So will the Continuous Integration.

It's already hard enough to get people to like Elm, I guess we don't need another hackish way to build, with a repo to clone and npm install and all that jazz. Doesn't look professionnal, and folks would use it as an argument (rightfully) against Elm.

**Comment by gdotdesign** [comment]
I'll try to work on this during the holidays, and If I can't, i'll release this under a tag something like `npm install elm-github-install@ssh-support -g`

File: src/index.js [hunk 2]
```diff
@@ -1,143 +1,270 @@
...
-}
+var getVersions = function (packageName) {
+
+    // console.log("getVersions: package=" + package);
+
+    return new Promise(function (fulfill, reject) {
+
+        var cmd;
+        if (isSsh(packageName)) {
+            cmd = 'git ls-remote git+ssh://' + repoHost(packageName) + '/' + packageName + ".git | awk -F/ '{ print $3 }'";
```
Comment (bbqbaron) [inline src/index.js]: Thank you so much for providing this functionality! In my local testing, I need to prepend `"git@"` to `repoHost(packageName)` in order to `ls-remote` a private Github package successfully. This seems to happen in the clone commands, so I think it's okay to do here, but I'm not a Git expert. +/-?

Comment (vankeisb) [reply src/index.js]: You are right : I have tested `git ls-remote` on various repos (private and public, github, ghe and bitbucket) and it works. I have pushed a fix.

**Comment by gdotdesign** [comment]
#7 supports this and the node.js version will be discontinued once it is merged, pre-releases are available in the releases page.

**Pull request closed without merging**